docs/api/api-config.md
<!-- ***********************************************************************************************
	Config API
************************************************************************************************ -->
# `Config` API {#config-api}

The `Config` object controls various aspects of SugarCube's behavior.

<p role="note"><b>Note:</b>
<code>Config</code> object settings should be placed within your project's JavaScript section (Twine&nbsp;2: the Story JavaScript; Twine&nbsp;1/Twee: a <code>script</code>-tagged passage).
</p>


<!-- ***************************************************************************
	General
**************************************************************************** -->
## General Settings<!-- legacy --><span id="config-api-miscellaneous"></span><!-- /legacy --> {#config-api-general}

<!-- *********************************************************************** -->

### `Config.addVisitedLinkClass` ↔ `boolean` (default: `false`) {#config-api-property-addvisitedlinkclass}

Determines whether the `link-visited` class is added to internal passage links that go to previously visited passages—i.e., the passage already exists within the story history.

<p role="note"><b>Note:</b>
You <em>must</em> provide your own styling for the <code>link-visited</code> class as none is provided by default.
</p>

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether links to visited passages have the class added.

#### Examples:

```javascript
Config.addVisitedLinkClass = true;
```

#### CSS styles:

You will also need to specify a `.link-visited` style that defines the properties visited links should have. For example:

```css
.link-visited {
	color: purple;
}
```

<!-- *********************************************************************** -->

### `Config.cleanupWikifierOutput` ↔ `boolean` (default: `false`) {#config-api-property-cleanupwikifieroutput}

Determines whether the output of the Wikifier is post-processed into more sane markup—i.e., where appropriate, it tries to transition the plethora of `<br>` elements into `<p>` elements.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether post-processing occurs.

#### Examples:

```javascript
Config.cleanupWikifierOutput = true;
```

<!-- *********************************************************************** -->

### `Config.debug` ↔ `boolean` (default: `false`) {#config-api-property-debug}

Indicates whether SugarCube is running in test mode, which enables debug views and various optional debugging errors and warnings.  See the [*Test Mode* guide](#guide-test-mode) for more information.

<p role="note"><b>Note:</b>
This setting is automatically set based on whether you're using a testing mode in a Twine compiler—i.e., <em>Test</em> mode in Twine&nbsp;2, <em>Test Play From Here</em> in Twine&nbsp;1, or the test mode option (<code>-t</code>, <code>--test</code>) in Tweego.  You may, however, forcibly enable it if you need to for some reason—e.g., if you're using another compiler, which doesn't offer a way to enable test mode.
</p>

<p role="note" class="see"><b>See Also:</b>
<a href="#config-api-property-enableoptionaldebugging"><code>Config.enableOptionalDebugging</code> setting</a>.
</p>

#### History:

* `v2.2.0`: Introduced.

#### Value:

A `boolean` value signifying whether SugarCube is running in test mode.

#### Examples:

##### Forcibly enabling test mode

```javascript
// Forcibly enable test mode.
Config.debug = true;
```

##### Check if test mode is enabled (in JavaScript)

```javascript
if (Config.debug) {
	// Do something debug related.
}
```

##### Check if test mode is enabled (in macros)

```
<<if Config.debug>>
	/* do something debug related */
<</if>>
```

<!-- *********************************************************************** -->

### `Config.enableOptionalDebugging` ↔ `boolean` (default: `false`) {#config-api-property-enableoptionaldebugging}

Determines whether various optional debugging errors and warnings are enabled outside of test mode.

<p role="note" class="see"><b>See Also:</b>
<a href="#config-api-property-debug"><code>Config.debug</code> setting</a>.
</p>

List of optional errors and warnings: *(not exhaustive)*

* [`<<if>>` macro](#macros-macro-if) assignment error.  If enabled, returns an error when the `=` assignment operator is used within its conditional—e.g., `<<if $suspect = "Bob">>`.  Does not flag other assignment operators.

#### History:

* `v2.37.0`: Introduced.

#### Value:

A `boolean` value signifying whether optional debugging errors and warnings are enabled.

#### Examples:

```javascript
Config.enableOptionalDebugging = true;
```

<!-- *********************************************************************** -->

### `Config.loadDelay` ↔ *integer* `number` (default: `0`) {#config-api-property-loaddelay}

Sets the integer delay (in milliseconds) before the loading screen is dismissed, once the document has signaled its readiness.  Not generally necessary, however, some browsers render slower than others and may need a little extra time to get a media-heavy page done.  This allows you to fine tune for those cases.

#### History:

* `v2.0.0`: Introduced.

#### Value:

An *integer* `number` value denoting the delay (in milliseconds) before the loading screen is dismissed.

#### Examples:

```javascript
// Delay the dismissal of the loading screen by 2000ms (2s)
Config.loadDelay = 2000;
```


<!-- ***************************************************************************
	Audio
**************************************************************************** -->
## Audio Settings {#config-api-audio}

<!-- *********************************************************************** -->

### `Config.audio.pauseOnFadeToZero` ↔ `boolean` (default: `true`) {#config-api-property-audio-pauseonfadetozero}

Determines whether the audio subsystem automatically pauses tracks that have been faded to `0` volume (silent).

#### History:

* `v2.28.0`: Introduced.

#### Value:

A `boolean` value signifying whether tracks that have faded to `0` are automatically paused.

#### Examples:

```javascript
Config.audio.pauseOnFadeToZero = false;
```

<!-- *********************************************************************** -->

### `Config.audio.preloadMetadata` ↔ `boolean` (default: `true`) {#config-api-property-audio-preloadmetadata}

Determines whether the audio subsystem attempts to preload track metadata—meaning information about the track (e.g., duration), not its audio frames.

<p role="note"><b>Note:</b>
It is unlikely that you will ever want to disable this setting.
</p>

#### History:

* `v2.28.0`: Introduced.

#### Value:

A `boolean` value signifying whether track metadata is preloaded.

#### Examples:

```javascript
Config.audio.preloadMetadata = false;
```


<!-- ***************************************************************************
	History
**************************************************************************** -->
## History Settings {#config-api-history}

<!-- *********************************************************************** -->

### `Config.history.controls` ↔ `boolean` (default: `true`) {#config-api-property-history-controls}

Determines whether the story's history controls (Backward, Jump To, & Forward buttons) are enabled within the UI bar.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether the story history controls are enabled.

#### Examples:

```javascript
Config.history.controls = false;
```

<!-- *********************************************************************** -->

### `Config.history.maxStates` ↔ *integer* `number` (default: `40`) {#config-api-property-history-maxstates}

Sets the maximum number of states (moments) to which the history is allowed to grow.  Should the history exceed the limit, states will be dropped from the past (oldest first).

<p role="note" class="tip"><b>Tip:</b>
For game-oriented projects, as opposed to more story-oriented interactive fiction, a setting of <code>1</code> is <strong><em>strongly recommended</em></strong>.
</p>

#### History:

* `v2.0.0`: Introduced.
* `v2.36.0`: Reduced the default to `40`.

#### Value:

An *integer* `number` value denoting how many moments are allowed within the history.

#### Examples:

```javascript
// Limit the history to a single moment (recommended for games)
Config.history.maxStates = 1;

// Limit the history to 25 moments
Config.history.maxStates = 25;
```


<!-- ***************************************************************************
	Macros
**************************************************************************** -->
## Macros Settings {#config-api-macros}

<!-- *********************************************************************** -->

### `Config.macros.maxLoopIterations` ↔ *integer* `number` (default: `1000`) {#config-api-property-macros-maxloopiterations}

Sets the maximum number of iterations allowed before the [`<<for>>` macro](#macros-macro-for) conditional forms are terminated with an error.

<p role="note"><b>Note:</b>
This setting exists to prevent a misconfigured loop from making the browser unresponsive.
</p>

#### History:

* `v2.0.0`: Introduced.

#### Value:

An *integer* `number` value denoting the total allowed number of conditional form loop iteration (per instance).

#### Examples:

```javascript
// Allow only 5000 iterations
Config.macros.maxLoopIterations = 5000;
```

<!-- *********************************************************************** -->

### `Config.macros.typeSkipKey` ↔ `string` (default: `" "`, space) {#config-api-property-macros-typeskipkey}

Sets the default [`KeyboardEvent.key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) value that causes the currently running [`<<type>>` macro](#macros-macro-type) instance to finish typing its content immediately.

#### History:

* `v2.33.1`: Introduced.

#### Value:

A `string` value denoting the keyboard key that causes `<<type>>` to immediately finish typing.

#### Examples:

```javascript
// Change the default skip key to Control (CTRL)
Config.macros.typeSkipKey = 'Control';
```

<!-- *********************************************************************** -->

### `Config.macros.typeVisitedPassages` ↔ `boolean` (default: `true`) {#config-api-property-macros-typevisitedpassages}

Determines whether the [`<<type>>` macro](#macros-macro-type) types out content on previously visited passages or simply outputs it immediately.

#### History:

* `v2.32.0`: Introduced.

#### Value:

A `boolean` value signifying whether `<<type>>` types out previously visited passages.

#### Examples:

```javascript
// Do not type on previously visited passages
Config.macros.typeVisitedPassages = false;
```

<!-- *********************************************************************** -->

### <span class="deprecated">`Config.macros.ifAssignmentError` ↔ `boolean` (default: `true`)</span> {#config-api-property-macros-ifassignmenterror}

<p role="note" class="warning"><b>Deprecated:</b>
This setting has been deprecated and should no longer be used.  See the <a href="#config-api-property-enableoptionaldebugging"><code>Config.enableOptionalDebugging</code></a> setting for its replacement.
</p>

#### History:

* `v2.0.0`: Introduced.
* `v2.37.0`: Deprecated in favor of the `Config.enableOptionalDebugging` setting.


<!-- ***************************************************************************
	Navigation
**************************************************************************** -->
## Navigation Settings {#config-api-navigation}

<!-- *********************************************************************** -->

### `Config.navigation.override` ↔ `Function` (default: *none*) {#config-api-property-navigation-override}

Allows the destination of passage navigation to be overridden.  The callback is passed one parameter, the original destination passage name.  If its return value is falsy, the override is cancelled and navigation to the original destination continues unperturbed.  If its return value is truthy, the override succeeds and that value is used as the new destination of the navigation.

#### History:

* `v2.13.0`: Introduced.

#### Value:

The callback value (`Function`).

#### Examples:

```javascript
Config.navigation.override = (destinationPassage) => {
	/* code that returns a passage name or a falsy value */
};
```

##### Based upon a story variable

```javascript
Config.navigation.override = (destinationPassage) => {
	var sv = State.variables;

	// If $health is less-than-or-equal-to 0, go to the "You Died" passage instead.
	if (sv.health <= 0) {
		return 'You Died';
	}
};
```


<!-- ***************************************************************************
	Passages
**************************************************************************** -->
## Passages Settings {#config-api-passages}

<!-- *********************************************************************** -->

### `Config.passages.displayTitles` ↔ `boolean` (default: `false`) {#config-api-property-passages-displaytitles}

Determines whether the current passage name is combined with the story's name within the browser tab's title bar.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `boolean` value signifying whether passage names are combined with the story's name within the title bar.

#### Examples:

```javascript
Config.passages.displayTitles = true;
```

<!-- *********************************************************************** -->

### `Config.passages.nobr` ↔ `boolean` (default: `false`) {#config-api-property-passages-nobr}

Determines whether rendering passages have their leading/trailing newlines removed and all remaining sequences of newlines replaced with single spaces before they're rendered.  Equivalent to including the [`nobr` special tag](#special-tag-nobr) on every passage.

<p role="note"><b>Note:</b>
Does not affect <code>script</code> or <code>stylesheet</code> tagged passages, for Twine&nbsp;1/Twee, or the Story JavaScript or Story Stylesheet sections, for Twine&nbsp;2.
</p>

#### History:

* `v2.19.0`: Introduced.

#### Value:

A `boolean` value signifying whether rendering passages have a global no-break enabled.

#### Examples:

```javascript
Config.passages.nobr = true;
```

<!-- *********************************************************************** -->

### `Config.passages.onProcess` ↔ `Function` (default: *none*) {#config-api-property-passages-onprocess}

Allows custom processing of passage text.  The function is invoked each time the [`<Passage>.processText()` method](#passage-api-prototype-method-processtext) is called.  It is passed an abbreviated version of the associated passage's [`Passage` instance](#passage-api)—containing only the `name`, `tags`, and `text` properties.  Its return value should be the post-processed text.

<p role="note"><b>Note:</b>
Does not affect <code>script</code> or <code>stylesheet</code> tagged passages, for Twine&nbsp;1/Twee, or the Story JavaScript or Story Stylesheet sections, for Twine&nbsp;2.
</p>

<div role="note"><b>Note:</b>
	<p>The function will be called just before the built-in no-break passage processing if you're also using that.</p>
	<p>See the <a href="#config-api-property-passages-nobr"><code>Config.passages.nobr</code> setting</a> and <a href="#special-tag-nobr"><code>nobr</code> special tag</a>.</p>
</div>

#### History:

* `v2.30.0`: Introduced.

#### Value:

A callback value (`Function`).

#### Examples:

```javascript
// Change instances of "cat" to "dog" in passages
Config.passages.onProcess = (p) => p.text.replace(/\bcat(s?)\b/g, 'dog$1');
```

<!-- *********************************************************************** -->

### `Config.passages.start` ↔ `string` (Twine&nbsp;2 default: *user-selected*; Twine&nbsp;1/Twee default: `"Start"`) {#config-api-property-passages-start}

Sets the starting passage name, the very first passage that will be displayed.

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `string` value denoting the starting passage name.

#### Examples:

```javascript
Config.passages.start = 'That Other Starting Passage';
```

<!-- *********************************************************************** -->

### `Config.passages.transitionOut` ↔ `string` | *integer* `number` (default: *none*) {#config-api-property-passages-transitionout}

Determines whether outgoing passage transitions are enabled.  Valid values are the name of the property being animated, which causes the outgoing passage element to be removed once that transition animation is complete, or an integer delay (in milliseconds), which causes the outgoing passage element to be removed once the delay has expired.  You will also need some CSS styles to make this work—examples given below.

<p role="note"><b>Note:</b>
If using an integer delay, ideally, it should probably be slightly longer than the outgoing transition delay that you intend to use—e.g., an additional 25ms or so should be sufficient.
</p>

#### History:

* `v2.0.0`: Introduced.

#### Value:

A `string` value denoting the CSS property being animated or an *integer* `number` value denoting the integer delay.

#### Examples:

```javascript
// Remove outgoing elements when their opacity animation ends
Config.passages.transitionOut = 'opacity';

// Remove outgoing elements after 1500ms (1.5s)
Config.passages.transitionOut = 1500;
```

#### CSS styles:

At the very least you will need to specify a `.passage-out` style that defines the transition's end state.  For example:

```css
.passage-out {
	opacity: 0;
}
```

That probably won't be very pleasing to the eye, however, so you will likely need several styles to make something that looks half-decent.  For example, the following will give you a basic crossfade:

```css
#passages {
	position: relative;
}
.passage {
	left: 0;
	position: absolute;
	top: 0;
	transition: opacity 1s ease;
}
.passage-out {
	opacity: 0;
}
```

<!-- *********************************************************************** -->

### <span class="deprecated">`Config.passages.descriptions` ↔ `boolean` | `Object` | `Function` (default: *none*)</span> {#config-api-property-passages-descriptions}

<p role="note" class="warning"><b>Deprecated:</b>
This setting has been deprecated and should no longer be used.  See the <a href="#config-api-property-saves-descriptions"><code>Config.saves.descriptions</code></a> setting for its replacement.
</p>

#### History:

* `v2.0.0`: Introduced.
* `v2.37.0`: Deprecated in favor of the `Config.saves.descriptions` setting.


<!-- ***************************************************************************
	Saves
**************************************************************************** -->
## Saves Settings {#config-api-saves}

<!-- *********************************************************************** -->


src/config.js
/***********************************************************************************************************************

	config.js

	Copyright © 2013–2025 Thomas Michael Edwards <thomasmedwards@gmail.com>. All rights reserved.
	Use of this source code is governed by a BSD 2-clause "Simplified" License, which may be found in the LICENSE file.

***********************************************************************************************************************/
/* global Save, State, Story, getTypeOf */

var Config = (() => { // eslint-disable-line no-unused-vars, no-var
	// General settings.
	let _addVisitedLinkClass     = false;
	let _cleanupWikifierOutput   = false;
	let _debug                   = false;
	let _enableOptionalDebugging = false;
	let _loadDelay               = 0;

	// Audio settings.
	let _audioPauseOnFadeToZero = true;
	let _audioPreloadMetadata   = true;

	// State history settings.
	let _historyControls  = true;
	let _historyMaxStates = 40;

	// Macros settings.
	let _macrosMaxLoopIterations   = 1000;
	let _macrosTypeSkipKey         = '\x20'; // Space
	let _macrosTypeVisitedPassages = true;

	// Navigation settings.
	let _navigationOverride;

	// Passages settings.
	let _passagesDisplayTitles = false;
	let _passagesNobr          = false;
	let _passagesStart; // Set by `Story.init()`
	let _passagesOnProcess;
	let _passagesTransitionOut;

	// Saves settings.
	let _savesDescriptions;
	let _savesId; // NOTE: Initially set by `Story.init()`.
	let _savesIsAllowed;
	let _savesMaxAuto      = 0;
	let _savesMaxSlot      = 8;
	let _savesMetadata;
	let _savesVersion;
	/* legacy */
	let _savesAutoload; // [DEPRECATED]
	/* /legacy */

	// UI settings.
	let _uiStowBarInitially    = 800;
	let _uiUpdateStoryElements = true;


	/*******************************************************************************
		Error Constants.
	*******************************************************************************/

	const errMacrosIfAssignmentErrorDeprecated = '[DEPRECATED] Config.macros.ifAssignmentError has been deprecated, see Config.enableOptionalDebugging instead';
	const errPassagesDescriptionsDeprecated    = '[DEPRECATED] Config.passages.descriptions has been deprecated, see Config.saves.descriptions instead';
	const errSavesAutoloadDeprecated           = '[DEPRECATED] Config.saves.autoload has been deprecated, see the Save.browser.continue API instead';
	const _baseSavesAutosaveDeprecated         = '[DEPRECATED] Config.saves.autosave has been deprecated';
	const errSavesOnLoadDeprecated             = '[DEPRECATED] Config.saves.onLoad has been deprecated, see the Save.onLoad API instead';
	const errSavesOnSaveDeprecated             = '[DEPRECATED] Config.saves.onSave has been deprecated, see the Save.onSave API instead';
	const errSavesSlotsDeprecated              = '[DEPRECATED] Config.saves.slots has been deprecated, see Config.saves.maxSlotSaves instead';
	const errSavesTryDiskOnMobileDeprecated    = '[DEPRECATED] Config.saves.tryDiskOnMobile has been deprecated';


	/*******************************************************************************
		Object Exports.
	*******************************************************************************/

	return Object.freeze({
		/*
			General settings.
		*/
		get addVisitedLinkClass() { return _addVisitedLinkClass; },
		set addVisitedLinkClass(value) { _addVisitedLinkClass = Boolean(value); },

		get cleanupWikifierOutput() { return _cleanupWikifierOutput; },
		set cleanupWikifierOutput(value) { _cleanupWikifierOutput = Boolean(value); },

		get debug() { return _debug; },
		set debug(value) { _debug = Boolean(value); },

		get enableOptionalDebugging() { return _enableOptionalDebugging; },
		set enableOptionalDebugging(value) { _enableOptionalDebugging = Boolean(value); },

		get loadDelay() { return _loadDelay; },
		set loadDelay(value) {
			if (!Number.isSafeInteger(value) || value < 0) {
				throw new RangeError('Config.loadDelay must be a non-negative integer number');
			}

			_loadDelay = value;
		},

		/*
			Audio settings.
		*/
		audio : Object.freeze({
			get pauseOnFadeToZero() { return _audioPauseOnFadeToZero; },
			set pauseOnFadeToZero(value) { _audioPauseOnFadeToZero = Boolean(value); },

			get preloadMetadata() { return _audioPreloadMetadata; },
			set preloadMetadata(value) { _audioPreloadMetadata = Boolean(value); }
		}),

		/*
			State history settings.
		*/
		history : Object.freeze({
			// TODO: (v3) This should be under UI settings → `Config.ui.historyControls`.
			get controls() { return _historyControls; },
			set controls(value) {
				const controls = Boolean(value);

				if (_historyMaxStates === 1 && controls) {
					throw new Error('Config.history.controls must be false when Config.history.maxStates is 1');
				}

				_historyControls = controls;
			},

			get maxStates() { return _historyMaxStates; },
			set maxStates(value) {
				if (!Number.isSafeInteger(value) || value < 1) {
					throw new RangeError('Config.history.maxStates must be a positive integer number');
				}

				_historyMaxStates = value;

				// Force `Config.history.controls` to `false`, when limited to `1` moment.
				if (_historyControls && value === 1) {
					_historyControls = false;
				}
			}
		}),

		/*
			Macros settings.
		*/
		macros : Object.freeze({
			get maxLoopIterations() { return _macrosMaxLoopIterations; },
			set maxLoopIterations(value) {
				if (!Number.isSafeInteger(value) || value < 1) {
					throw new RangeError('Config.macros.maxLoopIterations must be a positive integer number');
				}

				_macrosMaxLoopIterations = value;
			},

			get typeSkipKey() { return _macrosTypeSkipKey; },
			set typeSkipKey(value) { _macrosTypeSkipKey = String(value); },

			get typeVisitedPassages() { return _macrosTypeVisitedPassages; },
			set typeVisitedPassages(value) { _macrosTypeVisitedPassages = Boolean(value); },

			/* legacy */
			// Die if the deprecated macros if assignment error getter is accessed.
			get ifAssignmentError() { throw new Error(errMacrosIfAssignmentErrorDeprecated); },
			// Warn if the deprecated macros if assignment error setter is assigned to,
			// while also setting `Config.enableOptionalDebugging` for compatibilities sake.
			set ifAssignmentError(value) {
				console.warn(errMacrosIfAssignmentErrorDeprecated);
				Config.enableOptionalDebugging = value;
			}
			/* /legacy */
		}),

		/*
			Navigation settings.
		*/
		navigation : Object.freeze({
			get override() { return _navigationOverride; },
			set override(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.navigation.override must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				_navigationOverride = value;
			}
		}),

		/*
			Passages settings.
		*/
		passages : Object.freeze({
			// TODO: (v3) This should be under Navigation settings → `Config.navigation.updateTitle`.
			get displayTitles() { return _passagesDisplayTitles; },
			set displayTitles(value) { _passagesDisplayTitles = Boolean(value); },

			get nobr() { return _passagesNobr; },
			set nobr(value) { _passagesNobr = Boolean(value); },

			get onProcess() { return _passagesOnProcess; },
			set onProcess(value) {
				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (valueType !== 'function') {
						throw new TypeError(`Config.passages.onProcess must be a function, null, or undefined (received: ${valueType})`);
					}
				}

				_passagesOnProcess = value;
			},

			// TODO: (v3) This should be under Navigation settings → `Config.navigation.startPassageName`.
			get start() { return _passagesStart; },
			set start(value) {
				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (valueType !== 'string') {
						throw new TypeError(`Config.passages.start must be a string, null, or undefined (received: ${valueType})`);
					}
				}

				_passagesStart = value;
			},

			// TODO: (v3) This should be under Navigation settings → `Config.navigation.transitionOut`.
			get transitionOut() { return _passagesTransitionOut; },
			set transitionOut(value) {
				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (
						valueType !== 'string'
						&& (valueType !== 'number' || !Number.isSafeInteger(value) || value < 0)
					) {
						throw new TypeError(`Config.passages.transitionOut must be a string, non-negative integer number, null, or undefined (received: ${valueType})`);
					}
				}

				_passagesTransitionOut = value;
			},

			/* legacy */
			// Die if the deprecated passages descriptions getter is accessed.
			get descriptions() { throw new Error(errPassagesDescriptionsDeprecated); },
			// Warn if deprecated passages descriptions setter is assigned to,
			// then pass the value to the `Config.saves.descriptions` for
			// compatibilities sake.
			set descriptions(value) {
				console.warn(errPassagesDescriptionsDeprecated);

				switch (typeof value) {
					case 'boolean': {
						if (value && !Config.saves.descriptions) {
							Config.saves.descriptions = function () {
								return State.passage;
							};
						}

						break;
					}

					case 'function': {
						if (!Config.saves.descriptions) {
							Config.saves.descriptions = value;
						}

						break;
					}

					case 'undefined':
					case 'object': {
						if (value && !Config.saves.descriptions) {
							const dict = value;
							Config.saves.descriptions = function () {
								return Object.hasOwn(dict, State.passage) && dict[State.passage];
							};
						}

						break;
					}

					default:
						throw new TypeError(`Config.passages.descriptions must be a boolean, object, function, null, or undefined (received: ${getTypeOf(value)})`);
				}
			}
			/* /legacy */
		}),

		/*
			Saves settings.
		*/
		saves : Object.freeze({
			get descriptions() { return _savesDescriptions; },
			set descriptions(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.saves.descriptions must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				_savesDescriptions = value;
			},

			get id() { return _savesId; },
			set id(value) {
				if (typeof value !== 'string' || value === '') {
					throw new TypeError(`Config.saves.id must be a non-empty string (received: ${getTypeOf(value)})`);
				}

				_savesId = value;
			},

			get isAllowed() { return _savesIsAllowed; },
			set isAllowed(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.saves.isAllowed must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				_savesIsAllowed = value;
			},

			get maxAutoSaves() { return _savesMaxAuto; },
			set maxAutoSaves(value) {
				if (!Number.isSafeInteger(value)) {
					throw new TypeError('Config.saves.maxAutoSaves must be an integer number');
				}
				else if (value < 0 || value > Save.MAX_INDEX + 1) {
					throw new RangeError(`Config.saves.maxAutoSaves out-of-bounds (range: 0–${Save.MAX_INDEX + 1}; received: ${value})`);
				}

				_savesMaxAuto = value;
			},

			get maxSlotSaves() { return _savesMaxSlot; },
			set maxSlotSaves(value) {
				if (!Number.isSafeInteger(value)) {
					throw new TypeError('Config.saves.maxSlotSaves must be an integer number');
				}
				else if (value < 0 || value > Save.MAX_INDEX + 1) {
					throw new RangeError(`Config.saves.maxSlotSaves out-of-bounds (range: 0–${Save.MAX_INDEX + 1}; received: ${value})`);
				}

				_savesMaxSlot = value;
			},

			get metadata() { return _savesMetadata; },
			set metadata(value) {
				if (!(value == null || value instanceof Function)) { // nullish test
					throw new TypeError(`Config.saves.metadata must be a function, null, or undefined (received: ${getTypeOf(value)})`);
				}

				_savesMetadata = value;
			},

			get version() { return _savesVersion; },
			set version(value) { _savesVersion = value; },

			/* legacy */
			get _internal_autoload_() { // eslint-disable-line camelcase
				return _savesAutoload;
			},
			// Warn if the deprecated autoload getter is accessed.
			get autoload() {
				console.warn(errSavesAutoloadDeprecated);
				return _savesAutoload;
			},
			// Warn if the deprecated autoload setter is assigned to.
			set autoload(value) {
				console.warn(errSavesAutoloadDeprecated);

				if (value != null) { // nullish test
					const valueType = getTypeOf(value);

					if (
						valueType !== 'boolean'
						&& (valueType !== 'string' || value !== 'prompt')
						&& valueType !== 'function'
					) {
						throw new TypeError(`Config.saves.autoload must be a boolean, string ('prompt'), function, null, or undefined (received: ${valueType})`);
					}
				}

				_savesAutoload = value;
			},

			// Die if the deprecated saves autosave getter is accessed.
			get autosave() {
				throw new Error(`${_baseSavesAutosaveDeprecated}, see Config.saves.maxAutoSaves and Config.saves.isAllowed instead`);
			},
			// Die or warn if the deprecated saves autosave setter is assigned to,
			// while also setting `Config.saves.maxAutoSaves` and, possibly,
			// `Config.saves.isAllowed` for compatibilities sake.
			set autosave(value) {
				switch (typeof value) {
					case 'boolean':
						console.warn(`${_baseSavesAutosaveDeprecated}, for boolean usage see Config.saves.maxAutoSaves instead`);
						break;

					case 'function': {
						console.warn(`${_baseSavesAutosaveDeprecated}, for function usage see Config.saves.isAllowed instead`);

						if (!Config.saves.isAllowed) {
							const callback = value;
							Config.saves.isAllowed = function (saveType) {
								// Allow all other types while testing auto saves.
								return saveType !== Save.Type.Auto || callback(saveType);
							};
						}

						break;
					}

					default: {
						console.warn(`${_baseSavesAutosaveDeprecated}, for tag usage see Config.saves.isAllowed instead`);

						if (
							!(value instanceof Array)
							|| value.length === 0
							|| value.some(tag => typeof tag !== 'string')
						) {
							const valueType = getTypeOf(value);
							throw new TypeError(`Config.saves.autosave must be a boolean, Array<string>, function, null, or undefined (received: ${valueType}${valueType === 'Array' ? '<any>' : ''})`);
						}

						if (!Config.saves.isAllowed) {
							const userTags = value;
							Config.saves.isAllowed = function (saveType) {
								// Allow all other types while testing auto saves.
								return (
									saveType !== Save.Type.Auto
									|| userTags.includesAny(Story.get(State.passage).tags)
								);
							};
						}

						break;
					}
				}

				if (Config.saves.maxAutoSaves === 0) {
					Config.saves.maxAutoSaves = 1;
				}
			},

			// Die if the deprecated saves onLoad handler getter is accessed.
			get onLoad() { throw new Error(errSavesOnLoadDeprecated); },
			// Warn if the deprecated saves onLoad handler setter is assigned to, then
			// pass the handler to the `Save.onLoad` API for compatibilities sake.
			set onLoad(value) {
				console.warn(errSavesOnLoadDeprecated);
				Save.onLoad.add(value);
			},

			// Die if the deprecated saves onSave handler getter is accessed.
			get onSave() { throw new Error(errSavesOnSaveDeprecated); },
			// Warn if the deprecated saves onSave handler setter is assigned to, then
			// pass the handler to the `Save.onSave` API for compatibilities sake.
			set onSave(value) {
				console.warn(errSavesOnSaveDeprecated);
				Save.onSave.add(value);
			},

			// Die if the deprecated saves slots getter is accessed.
			get slots() { throw new Error(errSavesSlotsDeprecated); },
			// Warn if the deprecated saves slots setter is assigned to, then pass
			// the value to the `Config.saves.maxSlotSaves` for compatibilities
			// sake.
			set slots(value) {
				console.warn(errSavesSlotsDeprecated);
				Config.saves.maxSlotSaves = value;
			},

			// Warn if the deprecated saves tryDiskOnMobile getter is accessed, then
			// return `true`.
			get tryDiskOnMobile() {
				console.warn(errSavesTryDiskOnMobileDeprecated);
				return true;
			},
			// Warn if the deprecated saves tryDiskOnMobile setter is assigned to.
			set tryDiskOnMobile(value) { console.warn(errSavesTryDiskOnMobileDeprecated); }
			/* /legacy */
		}),

		/*
			UI settings.
		*/
		ui : Object.freeze({
			get stowBarInitially() { return _uiStowBarInitially; },
			set stowBarInitially(value) {
				const valueType = getTypeOf(value);

				if (
					valueType !== 'boolean'
					&& (valueType !== 'number' || !Number.isSafeInteger(value) || value < 0)
				) {
					throw new TypeError(`Config.ui.stowBarInitially must be a boolean or non-negative integer number (received: ${valueType})`);
				}

				_uiStowBarInitially = value;
			},

			get updateStoryElements() { return _uiUpdateStoryElements; },
			set updateStoryElements(value) { _uiUpdateStoryElements = Boolean(value); }
		})
	});
})();
